spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamApplicationDefinitionBuilder.java

```diff
@@ -85,6 +88,33 @@ else if (argument.getName().equalsIgnoreCase("outputType")) {
 					}
 				}
 			}
+			if (appNode.isLongLivedNonStreamApp()) {
+				builder.setApplicationType(ApplicationType.app);
+			}
+			else {
+				if (m == 0) {
+					if (sourceDestination == null) {
+						builder.setApplicationType(ApplicationType.source);
+					}
+					else {
+						if (appNodes.size() == 1 && sinkDestination == null) {
+							builder.setApplicationType(ApplicationType.sink);
+						}
+						else {
+							builder.setApplicationType(ApplicationType.processor);
+						}
+					}
+				}
+				else {
+					if (m < appNodes.size() - 1 || sinkDestination != null) {
+						builder.setApplicationType(ApplicationType.processor);
+					}
+					else {
+						builder.setApplicationType(ApplicationType.sink);
+					}
+
+				}
+			}
 			if (m > 0) {
```

Comment:
The destinations and groups should only be set if `!appNode.isLongLivedNonStreamApp()`